### PR TITLE
Dedicated S3 Proxy and parquet merge engine

### PR DESCRIPTION
**Pull request by pellicceama**

### CODE REVIEW

File: src/s3Proxy.ts [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+    },
+    region:
+      s3Config.url.includes('cloudflarestorage') ||
+      s3Config.url.includes('localhost')
+        ? 'auto'
+        : s3Config.url.split('.')[2],
+  });
+
+  try {
+    if (method === 'LIST') {
```
Comment (Rodri77) [inline src/s3Proxy.ts:65]: Should we do switch instead? add constants for the available methods or a type?

File: src/s3Proxy.ts [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+  DeleteObjectCommand,
+} from '@aws-sdk/client-s3';
+import { Upload } from '@aws-sdk/lib-storage';
+import initWasm, {
+  transformParquetStream,
+} from './lib/parquet-wasm/parquet_wasm';
+import binary from './lib/parquet-wasm/parquet_wasm_bg.wasm';
+
+function debugTempLog(...message: any) {
+  // console.debug(message);
```
Comment (Rodri77) [inline src/s3Proxy.ts:18]: This func is used below but does nothing is this expected for now? Should we remove?

File: src/s3Proxy.ts [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+  // Combine all file streams
+  const combinedStream = new ReadableStream({
+    async start(controller) {
+      for (const stream of fileStreams) {
+        const reader = stream.getReader();
+        while (true) {
+          const { done, value } = await reader.read();
+          if (done) break;
+          controller.enqueue(value);
```
Comment (Rodri77) [inline src/s3Proxy.ts:171]: Do you think we need more control here? is there a chance that we have the same value queued multiple times?

File: src/s3Proxy.ts [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+  // Sort all files by their key (which includes the timestamp)
+  allFiles.sort((a, b) => (a.Key || '').localeCompare(b.Key || ''));
+
+  return c.json({
+    ...response,
+    Contents: allFiles,
+  });
+}
+
+function groupFilesByDay(files: any[]): Record<string, any[]> {
```
Comment (Rodri77) [inline src/s3Proxy.ts:324]: Should we update the type ```Record<string, string>```?

Comment (Rodri77) [reply src/s3Proxy.ts:324]: or Record<string, Array<string>>

File: src/s3Proxy.ts [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+    const day = file.Key?.split('/').slice(0, 3).join('/'); // Assumes y=YYYY/m=MM/d=DD structure
+    if (day) {
+      acc[day] = acc[day] || [];
+      acc[day].push(file);
+    }
+    return acc;
+  }, {});
+}
+
+function constructVirtualMergedFiles(filesByDay: Record<string, any[]>): any[] {
```
Comment (Rodri77) [inline src/s3Proxy.ts:335]: It would be good if we prevent usage of any where possible

**Review by Rodri77 — COMMENTED** [review]
Left some comments/questions

File: src/s3Proxy.ts (hunk 2 above)
Comment (pellicceama) [reply src/s3Proxy.ts:18]: It's OK, this is a feature that will likely require some tweaks but I don't want it to log everything by default. In the near future we're going to want to implement a better global logger

File: src/s3Proxy.ts (hunk 3 above)
Comment (pellicceama) [reply src/s3Proxy.ts:171]: its ok, its just a bytes stream.

File: src/s3Proxy.ts (hunk 4 above)
Comment (pellicceama) [reply src/s3Proxy.ts:324]: its not strings though, its like a body.Contents object

File: src/s3Proxy.ts (hunk 5 above)
Comment (pellicceama) [reply src/s3Proxy.ts:335]: ditto above answer :)

**Pull request merged**